gzwrite.c

```diff
@@ -366,10 +366,17 @@ int ZEXPORT gzputs(file, str)
     if (state->mode != GZ_WRITE || state->err != Z_OK)
         return -1;
 
-    /* write string */
+    /* since an int is returned, make sure len fits in one, otherwise return
+       with an error (this avoids a flaw in the interface) */
     len = strlen(str);
-    ret = gz_write(state, str, len);
-    return ret == 0 && len != 0 ? -1 : ret;
+    if (len > INT_MAX) {
+        gz_error(state, Z_DATA_ERROR, "requested length does not fit in int");
+        return -1;
+    }
+
+    /* write len bytes from buf (the return value will fit in an int) */
+    ret = (int)gz_write(state, str, len);
+    return ret < len ? -1 : ret;
```

Comment:
The doc for gz_write says: 

"Write len bytes from buf to file.  Return the number of bytes written.  If the returned value is less than len, then there was an error."

the old implementation looked like it was trying to allow for partial writes by returning a fractional value.  Not sure if the fractional value or the reporting of errors is more import.  Fee free to ask me to change in either direction.